src/irmin-tezos-utils/parse.ml

```diff
@@ -0,0 +1,149 @@
+open Import
+open Files
+module Files = Make (Irmin_tezos.Conf) (Irmin_tezos.Schema)
+
+module Ring = struct
+  type t = { n : int; mutable i : int; a : int array }
+
+  let make n = { n; i = 0; a = Array.make n (-1) }
+
+  let add r x =
+    Array.set r.a r.i x;
+    r.i <- (r.i + 1) mod r.n
+
+  let get r i =
+    if i <= r.n then
+      let x = r.a.((r.i - i + r.n) mod r.n) in
+      if x <> -1 then Some x else None
+    else None
+end
+
+type ctx = { off : Int63.t; info : info }
+and info = { commit : int list; contents : int list; inode : int list }
+
+let info () = { commit = []; contents = []; inode = [] }
+
+type info_ring = { commit : Ring.t; contents : Ring.t; inode : Ring.t }
+
+let info_ring () =
+  { commit = Ring.make 1000; contents = Ring.make 1000; inode = Ring.make 1000 }
+
+let ctx_buffer = Bytes.create (4096 * 4096)
+
+let dump_ctx fd ctx =
+  let idx = ref 0 in
+  let f s =
+    String.iteri (fun i c -> Bytes.set ctx_buffer (!idx + i) c) s;
+    idx := !idx + String.length s
+  in
+  let flag =
+    List.length ctx.info.commit
+    + Int.shift_left (List.length ctx.info.contents) 2
+    + Int.shift_left (List.length ctx.info.inode) 4
+  in
+  Varint.encode_bin flag f;
+  List.iter (fun v -> Varint.encode_bin v f) ctx.info.commit;
+  List.iter (fun v -> Varint.encode_bin v f) ctx.info.contents;
+  List.iter (fun v -> Varint.encode_bin v f) ctx.info.inode;
+  let _ = Unix.write fd ctx_buffer 0 !idx in
+  (!idx, ctx.off)
+
+let dump_idx fd n is is2 =
+  let idx = ref 0 in
+  let f s =
+    String.iteri (fun i c -> Bytes.set ctx_buffer (!idx + i) c) s;
+    idx := !idx + String.length s
+  in
+  let off = List.fold_left (fun acc (i, _) -> i + acc) 0 is2 in
+  Varint.encode_bin (n + 1) f;
+  let _ =
+    List.fold_left2
+      (fun (i, i2) (i', off) (i2', _) ->
+        Varint.encode_bin i f;
+        Varint.encode_bin (i2 - i2') f;
+        Varint.encode_bin (Int63.to_int off) f;
+        (i + i', i2 - i2'))
+      (0, off) is is2
+  in
+  let _ = Unix.write fd ctx_buffer 0 !idx in
+  ()
+
+let get_values r = List.filter_map (Ring.get r) [ 1; 10; 1000 ]
+
+let main store_path info_last_path info_next_path idx_path =
+  let conf = Irmin_pack.Conf.init store_path in
+  match Files.File_manager.open_ro conf with
+  | Error exn -> Fmt.pr "%a\n%!" Files.Errs.pp exn
+  | Ok fm ->
+      let info_fd =
+        Unix.openfile info_last_path Unix.[ O_RDWR; O_CREAT; O_TRUNC ] 0o644
+      in
+      let len = ref (-1) in
+      let on_entry () _off _length _entry =
+        incr len;
+        ((), true)
+      in
+      let () = Files.fold fm ~on_entry () in
```

Comment:
can you maybe get the length from the size of the files on disk (instead of traversing the files)?

Reply:
The best solution would be to have access to this information via the file manager, as the idea of using was to never do any file system related actions ourselves.

This would also allow us to use `Append_only_file` for the files generated by the `parse function`.